resolver/SuggestPackage.cc

```diff
@@ -0,0 +1,135 @@
+#include "resolver/SuggestPackage.h"
+
+#include "common/formatting.h"
+#include "core/core.h"
+
+using namespace std;
+
+namespace sorbet::resolver {
+namespace {
+// Count the number of name parts in a symbol. Length should be same as `symbol2NameParts`.
+size_t nameLen(core::Context ctx, core::SymbolRef symbol) {
+    auto len = 0;
+    while (symbol.exists() && symbol != core::Symbols::root() && !symbol.data(ctx)->name.isPackagerName(ctx)) {
+        ++len;
+        symbol = symbol.data(ctx)->owner;
+    }
+    return len;
+}
+
+// Add all name parts for a symbol to a vector, exclude internal names used by packager.
+void symbol2NameParts(core::Context ctx, core::SymbolRef symbol, vector<core::NameRef> &out) {
+    ENFORCE(out.empty());
+    while (symbol.exists() && symbol != core::Symbols::root() && !symbol.data(ctx)->name.isPackagerName(ctx)) {
+        out.emplace_back(symbol.data(ctx)->name);
+        symbol = symbol.data(ctx)->owner;
+    }
+    std::reverse(out.begin(), out.end());
+}
+
+struct PackageMatch {
+    core::NameRef mangledName;
+};
+
+class PackageContext final {
+public:
+    core::Context ctx;
+    const core::packages::PackageInfo &currentPkg;
+
+    PackageContext(core::Context ctx) : ctx(ctx), currentPkg(ctx.state.packageDB().getPackageForFile(ctx, ctx.file)) {}
+
+    const core::packages::PackageDB &db() const {
+        return ctx.state.packageDB();
+    }
+
+    vector<PackageMatch> findPossibleMissingImports(const ast::ConstantLit::ResolutionScopes &scopes,
+                                                    core::NameRef name) const {
+        vector<PackageMatch> matches;
+        vector<core::NameRef> prefix;
+        // Resolution scopes are always longest to shortest. Reserve additional space for the name we're
+        // looking for.
+        prefix.reserve(nameLen(ctx, scopes.front()) + 1);
```

Comment:
This seems like a premature optimization. I'd remove this line and `nameLen`, and just let `prefix` resize naturally with the first run of `symbol2NameParts`. That reduces the duplicate code between `nameLen` and `symbol2NameParts`.